include/ghex/unstructured/communication_object_ipr.hpp

```diff
@@ -0,0 +1,573 @@
+/* 
+ * GridTools
+ * 
+ * Copyright (c) 2014-2020, ETH Zurich
+ * All rights reserved.
+ * 
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * 
+ */
+#ifndef INCLUDED_GHEX_UNSTRUCTURED_COMMUNICATION_OBJECT_IPR_HPP
+#define INCLUDED_GHEX_UNSTRUCTURED_COMMUNICATION_OBJECT_IPR_HPP
+
+// TO DO: remove unneeded includes
+#include "../communication_object_ipr.hpp"
+#include "./grid.hpp"
+#include "../pattern.hpp"
+#include "../cuda_utils/stream.hpp"
+#include "../packer.hpp"
+#include "../common/utils.hpp"
+#include "../common/test_eq.hpp"
+#include "../buffer_info.hpp"
+#include "../transport_layer/tags.hpp"
+#include "../arch_traits.hpp"
+#include <map>
+#include <unordered_map>
+#include <queue>
+#include <stdio.h>
+#include <functional>
+
+namespace gridtools {
+
+    namespace ghex {
+
+        // namespace unstructured {
+
+            /** @brief handle type for waiting on asynchronous communication processes.
+             * The wait function is stored in a member.
+             * @tparam Communicator communicator type
+             * @tparam Index index type for domain (local indices) and iteration space
+             * @tparam DomainIdType domain id type*/
+            template<typename Communicator, typename Index, typename DomainIdType>
+            class communication_handle_ipr<Communicator, unstructured::detail::grid<Index>, DomainIdType> {
+
+                private:
+
+                    // member types
+                    using communicator_type = Communicator;
+                    using index_type = Index;
+                    using grid_type = unstructured::detail::grid<index_type>;
+                    using domain_id_type = DomainIdType;
+
+                    // friend class // TO DO: is friend needed?
```

Comment:
the friend is needed because of private ctor

Reply:
Right. Done.